File: src/rust-axum/src/sql.rs
use std::env;
use std::path::Path;

pub fn load_sql(relative_path: &str) -> Result<String, std::io::Error> {
    let base_dir = if let Ok(queries_dir) = env::var("QUERIES_DIR") {
        Path::new(&queries_dir).to_path_buf()
    } else {
        // repo_root/database/queries
        Path::new(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .parent()
            .unwrap()
            .join("database")
            .join("queries")
    };
    
    let file_path = base_dir.join(relative_path);
    std::fs::read_to_string(file_path)
}

// SQL query constants - loaded at compile time for better performance

// Auth
pub const SQL_LOGIN: &str = include_str!("../../../database/queries/auth/login.sql");
pub const SQL_ME: &str = include_str!("../../../database/queries/auth/me.sql");

// Users
pub const SQL_CREATE_USER: &str = include_str!("../../../database/queries/users/create.sql");
pub const SQL_GET_USER: &str = include_str!("../../../database/queries/users/get.sql");
pub const SQL_LIST_USERS: &str = include_str!("../../../database/queries/users/list.sql");
pub const SQL_UPDATE_USER: &str = include_str!("../../../database/queries/users/update.sql");
pub const SQL_DELETE_USER: &str = include_str!("../../../database/queries/users/delete.sql");

// Posts
pub const SQL_CREATE_POST: &str = include_str!("../../../database/queries/posts/create.sql");
pub const SQL_LIST_POSTS: &str = include_str!("../../../database/queries/posts/list.sql");
pub const SQL_GET_POST: &str = include_str!("../../../database/queries/posts/get.sql");
pub const SQL_GET_POST_AUTHOR: &str = include_str!("../../../database/queries/posts/get_author.sql");
pub const SQL_DELETE_POST: &str = include_str!("../../../database/queries/posts/delete.sql");

// Comments
pub const SQL_CREATE_COMMENT: &str = include_str!("../../../database/queries/comments/create.sql");
pub const SQL_LIST_COMMENTS: &str = include_str!("../../../database/queries/comments/list.sql");

// Likes
pub const SQL_LIKE_EXISTS: &str = include_str!("../../../database/queries/likes/exists.sql");
pub const SQL_CREATE_LIKE: &str = include_str!("../../../database/queries/likes/create.sql");
pub const SQL_DELETE_LIKE: &str = include_str!("../../../database/queries/likes/delete.sql");

File: src/python-fastapi/src/sql.py
import os
from pathlib import Path


def load_sql(relative_path: str) -> str:
    base_dir_env = os.getenv("QUERIES_DIR")
    if base_dir_env:
        base_dir = Path(base_dir_env)
    else:
        # repo_root/database/queries
        base_dir = Path(__file__).resolve().parents[3] / "database" / "queries"
    file_path = base_dir / relative_path
    try:
        return file_path.read_text(encoding="utf-8")
    except FileNotFoundError as exc:
        raise RuntimeError(f"SQL file not found: {file_path}") from exc


# Auth
SQL_LOGIN = load_sql("auth/login.sql")
SQL_ME = load_sql("auth/me.sql")

# Users
SQL_CREATE_USER = load_sql("users/create.sql")
SQL_GET_USER = load_sql("users/get.sql")
SQL_LIST_USERS = load_sql("users/list.sql")
SQL_UPDATE_USER = load_sql("users/update.sql")
SQL_DELETE_USER = load_sql("users/delete.sql")

# Posts
SQL_CREATE_POST = load_sql("posts/create.sql")
SQL_LIST_POSTS = load_sql("posts/list.sql")
SQL_GET_POST = load_sql("posts/get.sql")
SQL_GET_POST_AUTHOR = load_sql("posts/get_author.sql")
SQL_DELETE_POST = load_sql("posts/delete.sql")

# Comments
SQL_CREATE_COMMENT = load_sql("comments/create.sql")
SQL_LIST_COMMENTS = load_sql("comments/list.sql")

# Likes
SQL_LIKE_EXISTS = load_sql("likes/exists.sql")
SQL_CREATE_LIKE = load_sql("likes/create.sql")
SQL_DELETE_LIKE = load_sql("likes/delete.sql")
